hivemind/p2p/p2p_daemon.py

```diff
@@ -315,38 +323,50 @@ async def stop_listening(self):
                 self._listen_task = None
                 self._server_stopped.clear()
 
-    async def add_stream_handler(self, name, handle):
+    async def add_stream_handler(self, name: str, handle: Callable[[bytes], bytes]) -> None:
         if self._listen_task is None:
             self.start_listening()
         await self._client.stream_handler(name, self._handle_stream(handle))
 
-    async def add_unary_handler(self, name, handle, in_proto_type, out_proto_type):
+    async def add_unary_handler(self, name: str, handle: Callable[[Any, P2PContext], Any],
+                                in_proto_type: type, out_proto_type: type) -> None:
         if self._listen_task is None:
             self.start_listening()
-        context = P2PContext(id=self.id, port=self._host_port, handle_name=name)
         await self._client.stream_handler(
-            name, P2P._handle_unary_stream(handle, context, in_proto_type, out_proto_type))
+            name, self._handle_unary_stream(handle, name, in_proto_type, out_proto_type))
 
-    async def call_peer_handler(self, peer_id, handler_name, input_data):
-        libp2p_peer_id = PeerID.from_base58(peer_id)
-        stream_info, reader, writer = await self._client.stream_open(libp2p_peer_id, (handler_name,))
+    async def call_peer_handler(self, peer_id: PeerID, handler_name: str, input_data: bytes) -> bytes:
+        stream_info, reader, writer = await self._client.stream_open(peer_id, (handler_name,))
         try:
             await P2P.send_raw_data(input_data, writer)
             return await P2P.receive_raw_data(reader)
         finally:
             writer.close()
 
+    async def call_unary_handler(self, peer_id: PeerID, handler_name: str,
+                                 request_protobuf: Any, response_proto_type: type) -> Any:
+        stream_info, reader, writer = await self._client.stream_open(peer_id, (handler_name,))
+        try:
+            await P2P.send_protobuf(request_protobuf, type(request_protobuf), writer)
+            result, err = await P2P.receive_protobuf(response_proto_type, reader)
+            if err is not None:
+                raise P2PHandlerError(f'Failed to call unary handler {handler_name} at {peer_id}: {err.message}')
+
+            return result
+        finally:
+            writer.close()
+
     def __del__(self):
         self._terminate()
 
     @property
-    def is_alive(self):
+    def is_alive(self) -> bool:
         return self._alive
 
-    async def shutdown(self):
+    async def shutdown(self) -> None:
         await asyncio.get_event_loop().run_in_executor(None, self._terminate)
 
-    def _terminate(self):
+    def _terminate(self) -> None:
         self._alive = False
         if self._child is not None and self._child.poll() is None:
             self._child.kill()
```

Comment:
nit: this method does not seem to have a good justification for SIGKILL. I would recommend using SIGTERM
```suggestion
            self._child.terminate()
```
The difference is that a SIGTERM gives the program a chance to close gracefully (closing files, network connections, freeing memory, etc), whereas SIGKILL doesn't.